Add: option to disable scrolling on miniconsoles

vadi2 opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Add an option to disable scrolling on miniconsoles
#### Motivation for adding to Mudlet
Address visible user pain point - some specialised miniconsoles don't benefit at all from being scrollable, in fact it is a detriment to be able to scroll them
#### Other info (issues closed, discussion etc)
Would help issues like https://github.com/svof/svof/issues/827 and this:
![image](https://github.com/Mudlet/Mudlet/assets/110988/62d57ebb-f709-45a1-b296-d090b95ce865)
https://discord.com/channels/283581582550237184/283582068334526464/1098816505484955679.

Scrolling can't be disabled on the main window, that's largely not an issue for players, and disabling scrolling there opens a can of worms.

API changes to test:
* `enableScrolling(window name)` - disables scrolling of a miniconsole
* `disableScrolling(window name)` - enables scrolling of a miniconsole

vadi2 committed:
Add option to disable scrolling on miniconsoles

SlySven commented on src/TConsole.cpp:
```diff
@@ -1069,6 +1069,10 @@ void TConsole::setConsoleBgColor(int r, int g, int b, int a)
 
 void TConsole::scrollDown(int lines)
 {
+    if (!mScrollingEnabled) {
```
```suggestion
    if ((mType & (UserWindow|SubConsole)) && !mScrollingEnabled) {
```

SlySven replied:
Oops, the same change is also desired by me on the `scrollUp(...)` method.

vadi2 replied:
Applied.

SlySven commented on src/TConsole.cpp:
```diff
@@ -1709,6 +1717,10 @@ void TConsole::setHorizontalScrollBar(bool isEnabled)
     }
 }
 
+void TConsole::setScrolling(bool enabled) {
+    mScrollingEnabled = enabled;
```
```suggestion
void TConsole::setScrolling(bool enabled)
{
    if (mType & (UserWindow|SubConsole)) {
        mScrollingEnabled = enabled;
    }
```

vadi2 replied:
Applied.

SlySven reviewed (commented):
Please confine the `TConsole`code to only work on the appropriate types!

Also - we should consider at least a visible indicator and/or a Mudlet event to be raised should the window's content extended beyond the bottom of the visible window so code that is painting text onto such a **scroll-less** window knows that it has overflowed!

vadi2 commented:
Good idea, I will put that in.

vadi2 committed:
Apply code review changes

vadi2 committed:
Return true for successful operation

vadi2 committed:
Merge branch 'development' into add-scrolling-option

demonnic committed:
Add handling for new enable/disableScrolling functions

demonnic commented:
ugh, I meant to do that last commit as a new branch with a PR to your branch but got distracted while working on it and forgot to make the new branch. Sorry for pushing directly into your PR like that.

I can pull it out and try again if you like. It just adds the necessary items to Geyser.MiniConsole

vadi2 commented:
All good with me, and my bad for forgetting to add it to Geyser at the same time

demonnic reviewed (approved):
LGTM. A little surprised it lets me approve now that I'm a joint author though.

SlySven reviewed (dismissed):
~~As per the second part of my comment https://github.com/Mudlet/Mudlet/pull/6848#pullrequestreview-1425398702 I have a PR to this PR that adds a `sysWindowOverflowEvent` with the window name and the number of lines that it has overflowed (scrolled off the top) as returned arguments.~~ Um, I accidentality merged it in after renaming a method as @vadi2 suggested. 

~~I was just about to post it up but~~ I have found one glitch in your current code in that if the normal scrolling is enabled and the window has been scrolled back and you then call `disableScrolling(....)` for it then the console does NOT scroll back down to the bottom - and the split gets locked into place - I think you need to do the equivalent of a `<Cntl>+<Enter>` on any window when `disableScrolling(....)` is applied to it...

SlySven committed:
Add overflow event for non scrolling consoles (#6872)

#### Brief overview of PR changes/additions
When a non-scrolling `TConsole` has text added to it that exceeds the
number of lines it can show this raises a `sysWindowOverflowEvent` with
the name of the console and the number of lines that have "scrolled off
the top" i.e. that have been hidden because there is not enough space
for them all. This can be used to manage the size of the text kept in
that window so that it is always fully visible.

#### Motivation for adding to Mudlet
This was prompted by the second part of my comment:
https://github.com/Mudlet/Mudlet/pull/6848#pullrequestreview-1425398702

#### Other issues
As I was looking at code that examines the number of lines of text in
the buffer I spotted that there were a couple of cases where the size of
the buffer or the size of a line in that buffer was being taken and then
being compared to be `<1` when really an `isEmpty()` test is a clearer
indication of intent - and where assignment to a temporary `(int)
length` variable could be made more streamlined.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>
Co-authored-by: Vadim Peretokin <vperetokin@hey.com>
Co-authored-by: eowmob <72653851+eowmob@users.noreply.github.com>

vadi2 commented:
> I think you need to do the equivalent of a <Cntl>+<Enter> on any window when disableScrolling(....) is applied to it...

Not sure, could be the desired behaviour? Would like to see if it's an issue for people using it and then update based on real world feedback.

SlySven committed:
Merge branch 'development' into add-scrolling-option

SlySven commented:
I've merged in the `development` branch so I can use the fix introduce by PR #6873 to hide some stuff (extra command lines!) whilst I demo the issue that I am currently moaning about here...

SlySven commented:
> > I think you need to do the equivalent of a + on any window when disableScrolling(....) is applied to it...
> 
> Not sure, could be the desired behaviour? Would like to see if it's an issue for people using it and then update based on real world feedback.

If the console is scrolled up when you disable scrolling the split gets locked on - look at the one on the right below.:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/76d22bf7-c2f6-422c-b6dc-810f4e2ddaf7)

 Worst, if you try to clear it, only the upper pane gets erased:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/cbfd110e-c27d-4949-b4b8-da0682e7b69e)

SlySven committed:
Merge branch 'development' into add-scrolling-option

vadi2 commented:
Yeah, the locking in place is intended right now - unless people find it inconvenient. It's not so obvious to me that everyone would want it scrolled down on lock. @demonnic your take?

Clearing of only the lower miniconsole is a bug that should be fixed :D

demonnic commented:
hmm. That's a tough one. I think in the end I since we already provide tools for turning scrolling off in code via `setScroll([windowName])` then letting it lock the scrolling where it is when scrolling is disabled is ok, as if they really want to they can use setScroll to make sure it's scrolled down first.

That being said I'd be fine if it did the equivalent of setScroll with no line number automatically either.

vadi2 commented:
OK, we're on the fence about it, so let's try @SlySven's suggestion. Approved in https://github.com/Mudlet/Mudlet/pull/6874#pullrequestreview-1453646890

SlySven committed:
Improve: clear split TConsole when disabling scrolling (#6874)

#### Brief overview of PR changes/additions
As I identified the code required was already used in two separate
places I refactored it out to a new method `(void)
TConsole::clearSplit()` so it could be used in a new, third place as
well.

#### Motivation for adding to Mudlet
As I mentioned in
https://github.com/Mudlet/Mudlet/pull/6848#issuecomment-1569192566
disabling scrolling when the `TConsole` had already been scrolled (and
had thus been split) locked the split in place and if the window was
cleared only the upper pane had all it's content removed. This addition
- as well as refactoring the code that clears the split from two
existing places - adds it to the code that disables scrolling so when
that happens any existing split is cleared.

#### Other info (issues closed, discussion etc)
DRY!

---------

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>
Co-authored-by: Vadim Peretokin <vperetokin@hey.com>
